doc/proposals/scorecard-custom-tests.md

```diff
@@ -0,0 +1,403 @@
+
+With this proposal, the scorecard will continue to setup and teardown the Operator being tested, as well as creating Custom Resources being tested.
+
+### User Stories
+
+#### Story 1
+
+Scorecard users will be able to construct custom tests and execute them using the scorecard, along with the built-in scorecard tests.
+
+#### Story 2
+
+Scorecard users can develop custom test using the programming language of their choice if that language offers similar support as golang client-go library.
+
+#### Story 3
+
+Scorecard users can execute custom tests as container images.
+
+## Design Details
+
+### Custom Test Definition
+
+* Tests that inspect the state of a Kube resource and compare that state against an expected value.  We’ll call these State Comparison tests.  State comparison tests might include a setup test step.  A test of this type would typically be implemented in a YAML manifest.
+
+* Tests that include logic, operator-specific logic, that determine if an operator is working as expected.  We’ll call these Operator Specific tests.   Operator-specific tests might include a setup test step.  A test of this type might be written in Java or Golang and packaged as a container image.
+
+* Tests that perform static checking of manifests.  We’ll call these Static Manifest tests.  A test of this type might be written in Java or Golang and packaged as a container image.  Static tests like this do not typically require a running Kube cluster but merely examine on-disk operator manifests for correctness.
+
+### Custom Test Components
+
+This proposal outlines the end-user facing components that make up a custom test including:
+
+#### Test Step
+
+A test step is part of a custom test and is meant to setup any Kube resources required for the specific test.  Test steps are coded in a YAML manifest and might include a Pod, Deployment, or Custom Resource manifest.  Test steps are read and processed by the scorecard in order to execute a custom test.  Test steps are read before a test assertion is executed by the scorecard.
+
+Test steps are expected to be provided by the scorecard end user in YAML format.  The path to these test steps would be configured within the scorecard configuration file or command line flags.
+
+#### Test Assertion
+
+A test assertion is the test check that is being performed.  The test assertion is written in a YAML manifest.  The scorecard executes test steps for a custom test prior to executing the test assertion.  The assertion is either true or false, or in the case of scorecard, a pass or fail.
+
+Test assertions are expected to be provided by the scorecard end user in YAML format.  The path to those assertions would be configured within the scorecard configuration file or command line flags.  Test assertions are related and physically organized on disk to match Test Steps.
+
```

Comment:
HI @jmccormick2001, 

Kudo has a diff design of SDK where for example all projects are managed by the same controller/manager. So, I was wonder if the Kutll design works with an SDK operator. Could you create a POC to see if we are able to do an Assert test/check with an SDK and Kuttll?

Reply:
I did a proof of concept, see the end of the proposal for details, that POC used by SDK generated operator (rqlite-operator) as the operator to test.  I then wrote Kudo tests that executed on that operator to prove that there was nothing required on the operator side to be able to run these sort of tests (e.g. no Kudo dependencies required).  I was able to construct 3 different tests to try it out.  The tests/code are public and referenced in the proposal for people to examine.  Interesting to note, that the goal of Kuttl is to detach this test capability from the larger Kudo project which I think makes sense for people just wanting to test their operator but not have to use Kudo for their operator development.